### DISCUSSION TITLE
Evolving Kubernetes for AI/ML

### DISCUSSION THREAD

**Discussion by johnbelamaric** [cncf/maintainer-summit#16] [Maintainer Summit EU 2025]
### Topic

Evolving Kubernetes for AI/ML

### What projects are involved?

Kubernetes, Kueue, Inference Gateway, JobSet, LeaderWorkSet

### Description

The community has many, many efforts ongoing to address the needs of AI/ML workloads. We think it would be a good idea to have a discussion of what areas are impacted, and what more we can or should be doing.

Here are some areas to understand, organized in sort of concentric circles starting at the node-level and moving up the stack. Some are related to AI/ML and some related simply to advancing infrastructure technologies (often in service of AI/ML). I am 100% sure I am missing many things below...

* Kubernetes Data Plane
  * Mutable container resources (in-place pod update)
  * Pod-level resources
  * Mutable nodes
  * Mutable containers in pods to support hierarchical scheduling
  * DRA / Device Management including CPU, GPU, NIC, memory
  * [OCI Volume Source](https://github.com/kubernetes/enhancements/issues/4639)
* Kubernetes Control Plane
  * Kube Scheduler (pod scheduling)
  * Workload level scheduling (bigger than pods)
  * Workload auto-scaling (including adapting to things like mutable pods)
  * Cluster auto-scaling (including adapting to things like mutable nodes)
  * Pod groups or similar resource reservation mechanisms
  * Supporting specialized schedulers
* Workload Controllers
  * JobSet
  * LeaderWorkSet
* Higher Layer Solutions
  * Kueue
  * Inference Gateway
* Ecosystem
  * Slurm
  * Ray
  * vLLM
  * Model serving / management
  * Kubeflow

Let's brainstorm!
* What other things does it impact?
* What else should we be doing?

### Session lead contact information

John Belamaric
@johnbelamaric

Propose also @mwielgus @ahg-g

**Comment by catblade** [comment]
Probably want to add PodGroup in here as well. Koordinator uses it, as does volcano (an equivalent, not the one in K8s proper)

**Comment by johnbelamaric** [comment]
It's under k8s control plane, but not sure it's the same thing

cc @dchen1107

**Comment by kerthcet** [comment]
I think model management should be part of it, the upstream effort like [OCI Volume Source](https://github.com/kubernetes/enhancements/issues/4639), which has been supported in multi runtimes, like containerd . Also should kubeflow part of the ecosystem, the kubernetes community has close integration with the project, like kueue, jobset etc. cc @andreyvelich @tenzen-y

**Reply by johnbelamaric** [reply]
added:
* K8s data plane: OCI volume source
* Ecosystem: model serving / management, Kubeflow

**Comment by andreyvelich** [comment]
This is great initiative @johnbelamaric, and thank you for the ping @kerthcet!
I agree that we should consolidate our efforts to improve Kubernetes for AI/ML workloads.

For instance, a key objective of [the Kubeflow Trainer V2 project](https://github.com/kubeflow/trainer/tree/master/docs/proposals/2170-kubeflow-training-v2#overview) (e.g. the next stage for Kubeflow Training Operator) was to leverage Kubernetes JobSet and Job as the foundation for our workloads. This approach helps to consolidate orchestration layers in a single operator, so we can focus on ML Experience + Additional Features on top (E.g. LLM Blueprints).

It would be also nice to enumerate Kubeflow project under Ecosystem or ML Lifecycle, given that this projects can be used as a standalone applications/workloads and Kubernetes-native nature for all Kubeflow projects (for reference Kubeflow projects in [the end-to-end ML lifecycle](https://www.kubeflow.org/docs/started/architecture/#kubeflow-components-in-the-ml-lifecycle)):


| Project | Role |
| --- | --- |
| Kubeflow Spark Operator | ML Feature Engineering/Data Processing |
| Kubeflow Trainer | ML Training and Fine-Tuning |
| Kubeflow Katib | ML Hyperparameter tuning and AutoML |
| Kubeflow KServe | ML Serving |
| Kubeflow Notebooks | ML Development Environment |
| Kubeflow Pipelines | ML Pipelines |


Please let me know what do you think!

cc @franciscojavierarceo @terrytangyuan @johnugeorge @kannon92 @bigsur0

**Comment by shivaycb** [comment]
Can we also include the Project KitOps during this discussion. KitOps is an open source DevOps tool that packages and versions your AI/ML model, datasets, code, and configuration into a reproducible artifact called a ModelKit. ModelKits are built on existing standards, ensuring compatibility with the tools your data scientists and developers already use.

and it was recently incorporated in the CNCF as a sandbox project = https://github.com/cncf/sandbox/issues/313

**Reply by johnbelamaric** [reply]
So, would that fit into ML lifecycle? I can make that a separate thing that "ecosystem" and add this and kubeflow? I need help with this taxonomy...

**Reply by shivaycb** [reply]
I think it is part of the ML lifecycle and compliments well with the notion for improving Kubernetes for AI/ML. Modelkits can be deployed as init containers in Kubernetes as a replacement for standard Docker containers, so they will support the full lifecycle of ML workloads on Kubernetes. But I do see your point regarding having a larger picture about the entire ecosystem of ML in Cloud native that includes kubeflow. KitOps stores everything as an OCI artifact so you can store it in any container registry (or use Jozu Hub which was designed for it)

**Reply by gorkem** [reply]
With KitOps and the accompanying "model-spec" effort we are trying to standardize an OCI Artifact definition for AI/ML artifacts. I think this is at least related with the "OCI Volume Source". I would love to have KitOps and the model-spec to be part of the "OCI Volume Source" conversation. Otherwise listing it as part of the "ecosystem" sounds good.

**Comment by CIngalls03** [comment]
This unconference session will take place at 13:50 - 14:25: https://sched.co/1w4xq

**Comment by lbocinco** [comment]
@johnbelamaric 
This Unconference topic was picked, congrats: https://maintainersummiteu2025.sched.com/event/1w4xq/unconference-session-2-evolving-kubernetes-for-aiml
